Fix #2251 : e2e test for parameters/expressions

giritheja opened the pull request:
Created parameters.js to perform e2e testing of parameters/expressions

arun1595 commented on core/tests/protractor/parameters.js:
```diff
@@ -0,0 +1,96 @@
+var editor = require('../protractor_utils/editor.js');
+var player = require('../protractor_utils/player.js');
+
+describe('Parameters', function() {
```
Can you add the file `protractor/parameters.js` to the `editorFeatures` suite in `protractor.conf.js`?

See https://github.com/oppia/oppia/blob/develop/core/tests/protractor.conf.js#L68

seanlip commented:
@giritheja, the e2e Travis checks are failing. Did you test this locally before you pushed?

arun1595 commented on core/tests/protractor.conf.js:
```diff
@@ -69,6 +69,7 @@ exports.config = {
       'protractor/gadgetEditor.js',
       'protractor/fallbacks.js',
       'protractor/historyTab.js'
```
@giritheja You're missing a comma at the end of the line.

giritheja commented:
@seanlip The tests after push did not show any failures. @arun1595 Are you facing same issues?

seanlip commented:
@giritheja -- see https://travis-ci.org/oppia/oppia/builds/182829786. These need to pass.

giritheja commented:
@seanlip I think the latest push should solve Travis issue. I had a little confusion on exploration level parameters, I have emailed @arun1595 about the same. I am waiting for his response before I can work on sorting that problem out.

seanlip commented:
Ah. Can you ask your question here, if it's easy to do so? It might be easier to get responses if discussions are public.

giritheja commented:
@seanlip @arun1595 I wanted a confirmation if I have to create a new method in the editor to create exploration level parameters. Since the current addParameters method of editor only creates card-level parameters.

seanlip commented:
Yes, please do that.

On Sunday, December 25, 2016, S Giritheja <notifications@github.com> wrote:

> @seanlip <https://github.com/seanlip> @arun1595
> <https://github.com/arun1595> I wanted a confirmation if I have to create
> a new method in the editor to create exploration level parameters. Since
> the current addParameters method of editor only creates card-level
> parameters.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/oppia/oppia/pull/2834#issuecomment-269092538>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AKFeyn-_0i-OHBhijdASOgO1WwcsieUxks5rLVNKgaJpZM4LJOtg>
> .
>

giritheja commented:
@arun1595 @seanlip The latest push addresses your comments, PTAL.

seanlip commented on core/tests/protractor_utils/editor.js:
```diff
@@ -397,6 +397,30 @@ var addParameterChange = function(paramName, paramValue) {
+    element(by.css('.protractor-test-add-param-button')).click();
+
+    var editorRowElem = element.all(by.css(
+    '.protractor-test-param-changes-list')).last();
```
Indent this continuation line by 2.

giritheja replied:
Corrected.

seanlip commented on core/tests/protractor_utils/editor.js:
```diff
@@ -397,6 +397,30 @@ var addParameterChange = function(paramName, paramValue) {
+    forms.AutocompleteDropdownEditor(editorRowElem).setValue(paramName);
+
+    /* Setting parameter value is difficult via css since the associated
+    input is a sub-component of the third party select2 library. We isolate
```
Are you sure about this? The parameter name is selected using a select2 dropdown, but the value is a standard input field, I think.

That said, this may be a reasonable approach anyway. I'll leave it to @arun1595 to make the final ruling (though you may still want to change the comment to reflect reality, if necessary).

giritheja replied:
I think I should probably wait for @arun1595 to comment on this.

arun1595 replied:
@giritheja Sean is right. The parameter name is controlled by `select2` but parameter value is a plain input field. So, please make sure to update your comment. The actual problem is the input fields are dynamically generated and they don't have classes defined, so there's no clean way of accessing the elements. 

Why not access the input field using `editorRowElem.all(by.tagName('input')).last()`? 
(I personally don't like the use of magic number 2 and one has to dig into the code to see why it is the third input field.)

What do you think?

arun1595 commented:
@giritheja LGTM except the review comment I left.

giritheja commented:
@seanlip @arun1595 I have updated the comments for addExplorationLevelParameterChange and addParameterChange. I have also changed the methods to find the last input field instead of using magic number 2 as suggested by Arun.
PTAL
